library/alloc/src/vec/drain.rs

```diff
@@ -129,15 +124,25 @@ impl<T, A: Allocator> Drop for Drain<'_, T, A> {
             }
         }
 
-        // exhaust self first
-        while let Some(item) = self.next() {
-            let guard = DropGuard(self);
-            drop(item);
-            mem::forget(guard);
-        }
+        let iter = mem::replace(&mut self.iter, (&mut []).iter());
+        let drop_len = iter.len();
+        let drop_offset = self.tail_start - iter.len();
```

Comment:
This assumes that all the yielded items are at the start of the range to drain, however since `Drain` implements `DoubleEndedIterator` they could be at the end too.